File: leetcode/c++/LCR_126_fei_bo_na_qi_shu_lie_lcof.cpp
// https://leetcode.cn/problems/fei-bo-na-qi-shu-lie-lcof

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int fib(int n) {
        if (n < 2)
            return n;
        const int mod = 1000000007;
        int dp[n+1];
        dp[0] =0, dp[1] = 1;
        for (int i = 2; i <= n; i++)
            dp[i] = (dp[i-1]  + dp[i-2]) % mod;
        return dp[n];
    }
};

// 时间复杂度为O(n)，空间复杂度为O(n)，可以优化为O(1)，见go版本

File: leetcode/golang/LCR_126_fei_bo_na_qi_shu_lie_lcof.go
func fib(n int) int {
    if n < 2 {
        return n
    }
    const mod = 1000000007
    a, b, sum := 0, 1, 0
    for i := 2; i <= n; i++ {
        sum = (a + b) % mod
        a = b
        b = sum
    }
    return sum 
}